### PR TITLE
Move to informer based controller for the CRD event handling

### PR DESCRIPTION
**Pull request by Praveenrajmani**

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+			klog.V(5).InfoS("Delete event", "key", key, "error", err)
+			queue.Add(Event{DeleteEvent, key, obj.(runtime.Object), nil})
+		},
+	})
+
+	return &Controller{
+		name:        name,
+		informer:    informer,
+		queue:       queue,
+		threadiness: 10,
```
Comment (Praveenrajmani) [inline pkg/controller/controller.go:129]: will accept threadiness as an argument

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	klog.Infof("Stopping %s controller", c.name)
+}
+
+// HasSynced is required for the cache.Controller interface.
+func (c *Controller) HasSynced() bool {
+	return c.informer.HasSynced()
+}
+
+// LastSyncResourceVersion is required for the cache.Controller interface.
+func (c *Controller) LastSyncResourceVersion() string {
```
Comment (balamurugana) [inline pkg/controller/controller.go:167]: Is it needed?

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+			klog.V(5).InfoS("Add event", "key", key, "error", err)
+			if err == nil {
+				queue.Add(Event{AddEvent, types.UID(key), obj.(runtime.Object), nil})
+			}
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldKey, oldErr := cache.MetaNamespaceKeyFunc(old)
+			newKey, newErr := cache.MetaNamespaceKeyFunc(new)
+			klog.V(5).InfoS("Update event", "oldKey", oldKey, "oldErr", oldErr, "newKey", newKey, "newErr", newErr)
+			queue.Add(Event{UpdateEvent, types.UID(newKey), new.(runtime.Object), old.(runtime.Object)})
```
Comment (balamurugana) [inline pkg/controller/controller.go:121]: Is it possible to have `oldKey != newKey`? if not, you could just use `oldKey`

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+func (c *Controller) opUnlock(lockKey types.UID) {
+	c.getOpLock(lockKey).Unlock()
+}
+
+func (c *Controller) processItem(ctx context.Context, event Event) error {
+	// Ensure that multiple operations on different versions of the same objects
+	// do not happen in parallel
+	c.opLock(event.Key)
+	defer c.opUnlock(event.Key)
```
Comment (balamurugana) [inline pkg/controller/controller.go:217]: You could rename to `lock()` and `unlock()`

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		handler.ObjectType(),
+		resyncPeriod,
+		cache.Indexers{},
+	)
+
+	queue := newRateLimitingQueue()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			klog.V(5).InfoS("Add event", "key", key, "error", err)
```
[the remark this answers is not in the document]
Comment (Praveenrajmani) [reply pkg/controller/controller.go:112]: the verbosity level in these logs are `5` which won't be printed by default. These logs might be useful for debugging purposes. Will change the log level to `8`.

File: pkg/controller/controller.go (hunk 2 above)
Comment (Praveenrajmani) [reply pkg/controller/controller.go]: no, removed

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	resyncPeriod   = 10 * time.Minute
```
Comment (balamurugana) [inline pkg/controller/controller.go:36]: This resync period affects drive discovery and initialization. We could accept this value as argument to `New()`. For drive discovery/initialization, we could have 5 minutes, others 10 minutes

**Review by balamurugana — COMMENTED** [review]
I tested controller code separately which is working good

File: pkg/controller/controller.go (hunk 6 above)
Comment (Praveenrajmani) [reply pkg/controller/controller.go]: done

**Review by balamurugana — APPROVED** [review]

**Pull request merged**